src/OrleansEventSourcing/EventStorage/LogViewAdaptor.cs

```diff
@@ -0,0 +1,360 @@
+        protected override async Task<int> WriteAsync()
+        {
+            var updates = GetCurrentBatchOfUpdates();
+
+            // record one of the guids in the batch, we will use it for idempotence check on failures
+            Guid guidForIdempotenceCheck = updates.First().Guid;
+
+            Services.Log(Severity.Verbose, "Write batch size={0} guid={1}", updates.Length, guidForIdempotenceCheck);
+
+            bool success;
+
+            try
+            {
+                success = await provider.EventStore.Append(
+                       streamName,
+                       updates.Select(u => new KeyValuePair<Guid,object>(u.Guid, u.Entry)),
+                       ConfirmedVersionInternal
+                    );
+
+                if (!success)
+                {
+                    LastPrimaryIssue.Record(new AppendEventsFailed() { IsOptimisticConcurrencyConflict = true }, Host, Services);
+                }
+                else
+                {
+                    LastPrimaryIssue.Resolve(Host, Services);
+
+                    Services.Log(Severity.Verbose, "Write successful");
+
+                    // update confirmed state
+                    foreach (var item in updates)
+                    {
+                        try
+                        {
+                            Host.UpdateView(ConfirmedStateInternal, item.Entry);
+                        }
+                        catch (Exception e)
+                        {
+                            Services.CaughtUserCodeException("UpdateView", nameof(WriteAsync), e);
+                        }
+
+                        ConfirmedVersionInternal++;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                success = false;
+                LastPrimaryIssue.Record(new AppendEventsFailed() { Exception = e, IsOptimisticConcurrencyConflict = false }, Host, Services);
+            }
+
+            if (!success)
+            {
+                Services.Log(Severity.Verbose, "Write apparently unsuccessful");
+
+                // an exception was thrown, or there was a concurrency failure.
+                // In either case, we need to read the latest state.
+                // if we see the guid in there, we know this append did actually succeeded
+                success = await ReadAsyncInternal(guidForIdempotenceCheck);
+
+                if (success)
+                    Services.Log(Severity.Verbose, "Write actually succeeded after all");
+            }
+
+            // let other clusters know that they should fetch the new events
+            // we may optimize this in the future by sending the events directly
+            if (success)
+                BroadcastNotification(new UpdateNotificationMessage()
+                {
+                    Version = ConfirmedVersionInternal,
+                    Updates = updates.Select(u => new KeyValuePair<Guid, TEvent>(u.Guid, u.Entry)).ToList(),
```

Comment:
hmm, this actually has some issues because some collections use TEvent, and some use object. I believe the cleanest way to avoid these problems is to make the IEventStorage.Load and IEventStorage.Append functions generic in the event type.